boa_engine/src/builtins/intl/tests.rs

```diff
@@ -0,0 +1,329 @@
+    let components = crate::builtins::intl::unicode_extension_components(&ext);
+    assert_eq!(components.attributes.is_empty(), true);
+    assert_eq!(components.keywords.len(), 3);
+    assert_eq!(components.keywords[0].key, "ca");
+    assert_eq!(components.keywords[0].value, "buddhist");
+    assert_eq!(components.keywords[1].key, "kk");
+    assert_eq!(components.keywords[1].value, "");
+    assert_eq!(components.keywords[2].key, "nu");
+    assert_eq!(components.keywords[2].value, "thai");
+
+    let ext = JsString::new("-u-ca-islamic-civil");
+    let components = crate::builtins::intl::unicode_extension_components(&ext);
+    assert_eq!(components.attributes.is_empty(), true);
+    assert_eq!(components.keywords.len(), 1);
+    assert_eq!(components.keywords[0].key, "ca");
+    assert_eq!(components.keywords[0].value, "islamic-civil");
+}
+
+#[test]
+fn locale_resolution() {
+    let mut context = Context::default();
+
+    // test lookup
+    let available_locales: Vec<JsString> = vec![];
+    let requested_locales: Vec<JsString> = vec![];
+    let relevant_extension_keys: Vec<JsString> = vec![];
+    let locale_data = FxHashMap::default();
+    let options = crate::builtins::intl::DateTimeFormatRecord {
+        locale_matcher: JsString::new("lookup"),
+        properties: FxHashMap::default(),
+    };
+
+    let locale_record = crate::builtins::intl::resolve_locale(
+        &available_locales,
+        &requested_locales,
+        &options,
+        &relevant_extension_keys,
+        &locale_data,
+        &mut context,
+    );
+    assert_eq!(
+        locale_record.locale,
+        crate::builtins::intl::default_locale()
+    );
+    assert_eq!(
+        locale_record.data_locale,
+        crate::builtins::intl::default_locale()
+    );
+    assert_eq!(locale_record.properties.is_empty(), true);
+
+    // test best fit
+    let available_locales: Vec<JsString> = vec![];
+    let requested_locales: Vec<JsString> = vec![];
+    let relevant_extension_keys: Vec<JsString> = vec![];
+    let locale_data = FxHashMap::default();
+    let options = crate::builtins::intl::DateTimeFormatRecord {
+        locale_matcher: JsString::new("best-fit"),
+        properties: FxHashMap::default(),
+    };
+
+    let locale_record = crate::builtins::intl::resolve_locale(
+        &available_locales,
+        &requested_locales,
+        &options,
+        &relevant_extension_keys,
+        &locale_data,
+        &mut context,
+    );
+    assert_eq!(
+        locale_record.locale,
+        crate::builtins::intl::default_locale()
+    );
+    assert_eq!(
+        locale_record.data_locale,
+        crate::builtins::intl::default_locale()
+    );
+    assert_eq!(locale_record.properties.is_empty(), true);
+
+    // available: [es-ES], requested: [es-ES]
+    let available_locales = vec![JsString::new("es-ES")];
+    let requested_locales = vec![JsString::new("es-ES")];
+    let relevant_extension_keys: Vec<JsString> = vec![];
```

Comment:
While they are the same, macros require expansion, so they increase compile times. Also, using the generic in the function call reduces the duplication of code :)

```suggestion
    let relevant_extension_keys = Vec::<JsString>::new();
```